fast_llm/data/data/gpt/data.py

```
@@ -154,9 +168,19 @@ def get_iterator(
         Assert.incl(dataset_name, self._datasets)
         Assert.in_range_incl(batch_config.sequence_length, 1, self._max_sequence_length)
         log_main_rank(f"Initializing {dataset_name} dataset iterator from sample {consumed_samples}...")
+
+        dataset = self._datasets[dataset_name]  # noqa
+
+        if hasattr(dataset._dataset, "_indexed_dataset") and hasattr(
```

This won't work, the actual dataset may be wrapped by concatenations, blending, splitting, etc., and there may be multiple base datasets with variable `_has_preference_spans`. Also we don't know if the model even wants spans.

Best option is to add its as an argument to `GPTData.__init__`. Then it can be forwarded to the dataset sampling through `GPTSamplingData`, so they can pick the right sampling method, i.e. skip loading spans if not needed and raise an error if we want spans that aren't there.

---

Thanks for sharing this @jlamypoirier - just wanted to confirm the general flow with you before implementing the changes

From my understanding we'd pass in a user parameter that would populate `GPTSamplingParameters`, this will help control the logic of whether we are using preference spans and how we're reading memmapped data.

There's also the other aspect where certain training objective require preference loss spans while others might not be compatible.  In this case, we would want to raise an error early when preference loss spans are required but not specified but skip span loading otherwise.

One difference between preference spans and regular loss masking spans is currently if a dataset is written with preference spans, it assumes it is being used since there are both chosen and rejected turns.  We'd have to define what skipping means - do we apply loss mask on the chosen turn or on the entire conversation or should we also just throw an error if a preference dataset is provided to a non-compatible training objective?

---

Ok, raising an error makes sense. 

(The disjoint use case worries me a bit, it looks like an entirely different dataset type rather than a new feature, so separate class(es) would be more appropriate. Llet's ignore for now since the infrastructure isn't ready, but keep it in mind for the future.)

---

Yeah makes sense since it's an entirely new training stage/objective.  For now, I've added an early check within config validation to make sure if dpo loss is enabled, preference pairs must be enabled too (and vice versa).  Memmap class also ensures that the dataset actually has these preference pairs or else an error is thrown.